ktor-server/ktor-server-host-common/jvm/src/io/ktor/server/engine/BaseApplicationResponse.kt

```diff
@@ -118,13 +118,17 @@ abstract class BaseApplicationResponse(override val call: ApplicationCall) : App
                 return respondWriteChannelContent(content)
             }
 
-            // Pipe is least efficient
+            // Pipe is the least efficient
             is OutgoingContent.ReadChannelContent -> {
                 // First call user code to acquire read channel, because it could fail
                 val readChannel = content.readFrom()
-                // If channel is fine, commit headers and pipe data
-                commitHeaders(content)
-                return respondFromChannel(readChannel)
+                try {
```

Comment:
Why **io.ktor.utils.io.ByteReadChannel** does not extends **io.ktor.utils.io.core.Closeable**?
Then we could use try-with-resource statement (**use** for kotlin).

Otherwise cancel() execute close() with kotlinx.coroutines.CancellationException but the reasons for close may be different.

P.S. Its just a question Im interested of patterns used in ktor project.

Reply:
ReadChannel is similar to ReceiveChannel that have cancel, not close. This doesn't prevent us from having a specialized use function that does cancel. We probably will do in the new kotlinx.io.

Reply:
Ok. Thanks for reply!)